typed/remove_test.go

```diff
@@ -0,0 +1,340 @@
+package typed
+
+import (
+	"fmt"
+	"testing"
+
+	"sigs.k8s.io/structured-merge-diff/v3/fieldpath"
+	"sigs.k8s.io/structured-merge-diff/v3/value"
+)
+
+func TestRemoveDeduced(t *testing.T) {
+
+	var cases = []struct {
+		object YAMLObject
+		remove *fieldpath.Set
+		expect YAMLObject
+	}{
+		{
+			object: `{}`,
+			remove: fieldpath.NewSet(fieldpath.MakePathOrDie("a")),
+			expect: `{}`,
+		},
+		{
+			object: `{"a": "value"}`,
+			remove: fieldpath.NewSet(fieldpath.MakePathOrDie("a")),
+			expect: ``,
+		},
+		{
+			object: `{"a": "value", "b": "value"}`,
+			remove: fieldpath.NewSet(fieldpath.MakePathOrDie("a")),
+			expect: `{"b": "value"}`,
+		},
+		{
+			object: `{"a": "value", "b": {}}`,
+			remove: fieldpath.NewSet(fieldpath.MakePathOrDie("a")),
+			expect: `{"b": {}}`,
+		},
+		{
+			object: `{"a": "value", "b": {"c":"value"}}`,
+			remove: fieldpath.NewSet(fieldpath.MakePathOrDie("b")),
+			expect: `{"a": "value"}`,
+		},
+		{
+			// this is okay, because in a deduced schema `b` is handled as a map not a struct
+			object: `{"a": "value", "b":{"c":"value"}}`,
+			remove: fieldpath.NewSet(fieldpath.MakePathOrDie("b")),
+			expect: `{"a": "value"}`,
+		},
+		{
+			object: `{"a": "value", "b": {"c":"value"}}`,
+			remove: fieldpath.NewSet(fieldpath.MakePathOrDie("b", "c")),
+			expect: `{"a": "value", "b": null}`,
+		},
+		{
+			object: `{"a": "value", "b": []}`,
+			remove: fieldpath.NewSet(fieldpath.MakePathOrDie("b")),
+			expect: `{"a": "value"}`,
+		},
+		{
+			object: `{"a": "value", "b": ["c"]}`,
+			remove: fieldpath.NewSet(fieldpath.MakePathOrDie("b", "c")),
+			// TODO: this should fail
+			expect: `{"a": "value", "b": ["c"]}`,
+		},
+		{
+			object: `{"a": "value", "b": ["c", "d"]}`,
+			remove: fieldpath.NewSet(fieldpath.MakePathOrDie("b", "c")),
+			expect: `{"a": "value", "b": ["c", "d"]}`,
+		},
+		{
+			object: `{"a": "value", "b": [{"c": "value"}, {"d": "value"}]}`,
+			remove: fieldpath.NewSet(fieldpath.MakePathOrDie("b", "c")),
+			expect: `{"a": "value", "b": [{"c": "value"}, {"d": "value"}]}`,
```

Comment:
I would expect
```
expect: `{"a": "value", "b": [{"d": "value"}]}`
```

Reply:
Same, the remove path has to be different for this. At least that's how Jenny explained it to me.
But I can add another test for this.

Reply:
OK, I agree that I'm wrong on this one.

Reply:
c is the key of an object in the list, it wouldn't make sense to remove "c" from all the objects in the list I think.